internal/expr/parser.go

```diff
@@ -716,30 +742,150 @@ func (p *Parser) parseOutputExpr() (*outputExpr, bool, error) {
 	return nil, false, nil
 }
 
-// parseInputExpr parses an input expression of the form "$Type.name".
-func (p *Parser) parseInputExpr() (expression, bool, error) {
+// parseSliceInputExpr parses an input expression of the form "$Type[:]".
+func (p *Parser) parseSliceInputExpr() (*sliceInputExpr, bool, error) {
 	cp := p.save()
 	if !p.skipByte('$') {
 		return nil, false, nil
 	}
 
-	// Case 1: Slice range, "Type[:]".
 	if st, ok, err := p.parseSliceAccessor(); err != nil {
+		cp.restore()
 		return nil, false, err
 	} else if ok {
 		return &sliceInputExpr{sliceTypeName: st, raw: p.input[cp.pos:p.pos]}, true, nil
 	}
 
-	// Case 2: Struct or map, "Type.something".
-	if ma, ok, err := p.parseTypeAndMember(); ok {
-		if ma.memberName == "*" {
-			return nil, false, errorAt(fmt.Errorf("asterisk not allowed in input expression %q", "$"+ma.String()), cp.lineNum, cp.colNum(), p.input)
+	cp.restore()
+	return nil, false, nil
+}
+
+// parseMemberInputExpr parses an input expression of the form "$Type.member".
+func (p *Parser) parseMemberInputExpr() (*memberInputExpr, bool, error) {
+	cp := p.save()
+	ma, ok, err := p.parseInputMemberAccessor()
+	if !ok {
```

Comment:
This is relying on the fact that `ok = false` when `err != nil`. I think in Go the convention is to check the `err` value, then the `ok` value. There are some functions in the standard library that do not return the zero value when returning an error for example.

Reply:
Right, but here we don't actually check the err value, if `ok=true` then its one of our invariant that `err=nil`. If we did check the err here then we would just be doing `if err != nil {return err}; if !ok {return nil} which has the same effect.

Can change it though if it makes you uncomfortable.

Reply:
What happens if there is an error in the parse function, we are checking `ok` and that might be set because the invariants only mandate `err != nil`. That is what I meant. We would conitnue with the function as normal instead of returning.